tap_appstore/client.py

```diff
@@ -18,12 +23,103 @@
 
 _Auth = Callable[[requests.PreparedRequest], requests.PreparedRequest]
 
-# TODO: Delete this is if not using json files for schema definition
-SCHEMAS_DIR = importlib_resources.files(__package__) / "schemas"
-
+logger = logging.getLogger(__name__)
 
 class AppStoreStream(RESTStream):
     """AppStore stream class."""
+    DATE_FORMAT = '%Y-%m-%d'
+    DATE_INCREMENT = timedelta(days=1)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.api = self.setup_api_connection()
+        self.float_fields = None
+        self.int_fields = None
+        self.date_fields = None
+
+    def setup_api_connection(self):
+        """Set up the API connection using provided configuration."""
+        return Api(self.config['key_id'], self.config['key_file'], self.config['issuer_id'], submit_stats=False)
+
+    def download_data(self, start_date, api):
+        """Set up the endpoint for the API call. Override in subclass as needed."""
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def process_record(self, record):
+        """Convert ints, floats, and date fields"""
+        if self.float_fields:
+            self.convert_fields(record, self.float_fields, float)
+
+        if self.int_fields:
+            self.convert_fields(record, self.int_fields, int)
+
+        if self.date_fields:
+            for field, format in self.date_fields.items():
+                if field in record and record[field]:
+                    record[field] = self.convert_date(record[field], format)
+
+        return record
+
+    def get_records(self, context: dict = None):
+        """Return a generator of record-type dictionary objects."""
+        line_id = 0
+        starting_timestamp = self.get_starting_timestamp(context)
+        start_date = starting_timestamp if starting_timestamp else self.config.get('start_date', '2024-04-01')
+        start_date = (start_date + self.DATE_INCREMENT).strftime(self.DATE_FORMAT)
```

Comment:
@joaopamaral We no longer manually update the state and instead use replication keys, though I am still working on making them work for `FinancialReportStream`.

However, even for the other Streams, I see the correct `replication_key_value`s being updated on each run:

```
2024-05-16T05:48:48.062474Z [info     ] 2024-05-16 11:18:48,062 | INFO     | target-parquet       | Emitting completed target state {"bookmarks": {"sales_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}, "subscriber_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}, "subscription_event_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}, "subscription_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}}} cmd_type=loader name=target-parquet run_id=bfda6f8c-19b0-45e0-ade7-3da306af7968 state_id=mahangu_test_22 stdio=stderr
```

And the state is correctly updated:

```
$ meltano state get mahangu_test_22
2024-05-16T05:49:31.193849Z [info     ] The default environment 'test' will be ignored for `meltano state`. To configure a specific environment, please use the option `--environment=<environment name>`.
{"singer_state": {"bookmarks": {"sales_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}, "subscriber_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}, "subscription_event_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}, "subscription_reports": {"replication_key": "start_date", "replication_key_value": "2024-04-03"}}}}
```

However, when running ` meltano el tap-appstore target-parquet --state-id mahangu_test_22` it doesn't incrementally keep updating until today.  As far as I can see from - https://sdk.meltano.com/en/latest/incremental_replication.html - specifying a `replication_key` in each child Class/Stream should trigger an incremental update?

Reply:
> However, when running  meltano el tap-appstore target-parquet --state-id mahangu_test_305 it doesn't incrementally keep updating until today but rather just stops after pulling data for just one run.

If get all records until today, there is no report left so it will not extract any data. Try to run again tomorrow.

> As far as I can see from - https://sdk.meltano.com/en/latest/incremental_replication.html - specifying a replication_key in each child Class/Stream should trigger an incremental update?

Yes!

> Note: a TODO is to not increment the date here if it's a first run. Maybe we should increment at the end of this function?

We will not store the incremented state, only the last extracted state that will be updated by singer sdk.

Reply:
> If get all records until today, there is no report left so it will not extract any data. Try to run again tomorrow.

@joaopamaral What I mean is, if I run it with a start date of `2024-04-01`, it only extracts data for April 1. I have to run it again to extract data for April 2. We can look into this further in our pairing session.